# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1786745955 -7200
# Parent  19b508e5a8a0746cf898070bb952b3c72b07dca3
Feature: Linux systray: Adding activation for Xfce, from
https://github.com/AyatanaIndicators/libayatana-appindicator/pull/71#issuecomment-1704357887:

Work by Gaël Bonithon (Tamaranch) <gael@xfce.org>

Ported to https://github.com/AyatanaIndicators/libayatana-appindicator-glib by DeekSeek.

diff --git a/third_party/appindicator/ayatana-appindicator.c b/third_party/appindicator/ayatana-appindicator.c
--- a/third_party/appindicator/ayatana-appindicator.c
+++ b/third_party/appindicator/ayatana-appindicator.c
@@ -55,8 +55,9 @@ enum
     PROP_LABEL_GUIDE,
     PROP_ORDERING_INDEX,
     PROP_TITLE,
     PROP_MENU,
+    PROP_ITEM_IS_MENU,
     PROP_ACTIONS,
     PROP_TOOLTIP_ICON,
     PROP_TOOLTIP_TITLE,
     PROP_TOOLTIP_DESCRIPTION
@@ -88,8 +89,9 @@ typedef struct _AppIndicatorPrivate
     gchar *sAccessibleDesc;
     gchar *sAttAccessibleDesc;
     guint nLabelChangeIdle;
     GMenu *pMenu;
+    gboolean bItemIsMenu;
     GSimpleActionGroup *pActions;
     guint nActionsId;
     guint nMenuId;
     GDBusConnection *pConnection;
@@ -391,8 +393,17 @@ static GVariant* onBusGetProperty (GDBus
 
         return pValue;
     }
 
+    nCompare = g_strcmp0 (sProperty, "ItemIsMenu");
+
+    if (nCompare == 0)
+    {
+        GVariant *pValue = g_variant_new_boolean (pPrivate->bItemIsMenu);
+
+        return pValue;
+    }
+
     nCompare = g_strcmp0 (sProperty, "XAyatanaLabel");
 
     if (nCompare == 0)
     {
@@ -1023,8 +1034,14 @@ static void app_indicator_set_property (
             pPrivate->pMenu = G_MENU (g_value_dup_object (pValue));
 
             break;
         }
+        case PROP_ITEM_IS_MENU:
+        {
+            pPrivate->bItemIsMenu = g_value_get_boolean (pValue);
+
+            break;
+        }
         case PROP_ACTIONS:
         {
             g_clear_object (&pPrivate->pActions);
             pPrivate->pActions = G_SIMPLE_ACTION_GROUP (g_value_dup_object (pValue));
@@ -1167,8 +1184,14 @@ static void app_indicator_get_property (
             g_value_set_object (pValue, pPrivate->pMenu);
 
             break;
         }
+        case PROP_ITEM_IS_MENU:
+        {
+            g_value_set_boolean (pValue, pPrivate->bItemIsMenu);
+
+            break;
+        }
         case PROP_ACTIONS:
         {
             g_value_set_object (pValue, pPrivate->pActions);
 
@@ -1352,8 +1375,16 @@ static void app_indicator_class_init (Ap
     GParamSpec *pSpecMenu = g_param_spec_string ("menu", "The menu of the application indicator", "The menu that should be shown when the Application Indicator is clicked on in the panel.", NULL, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS);
     g_object_class_install_property (pClass, PROP_MENU, pSpecMenu);
 
     /**
+     * AppIndicator:item-is-menu:
+     *
+     * Whether the application indicator only supports the context menu.
+     */
+    GParamSpec *pSpecItemIsMenu = g_param_spec_boolean ("item-is-menu", "Whether the application indicator only supports the context menu", "If TRUE the visualization should prefer showing the menu or sending ContextMenu() instead of Activate().", FALSE, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS);
+    g_object_class_install_property (pClass, PROP_ITEM_IS_MENU, pSpecItemIsMenu);
+
+    /**
      * AppIndicator:actions:
      *
      * The action group that is associated with the menu items.
      */
diff --git a/third_party/appindicator/notification-item.xml.c b/third_party/appindicator/notification-item.xml.c
--- a/third_party/appindicator/notification-item.xml.c
+++ b/third_party/appindicator/notification-item.xml.c
@@ -11,8 +11,9 @@ const char * _notification_item =
 "        <property name=\"AttentionAccessibleDesc\" type=\"s\" access=\"read\" />\n"
 "        <property name=\"Title\" type=\"s\" access=\"read\" />\n"
 "        <property name=\"IconThemePath\" type=\"s\" access=\"read\" />\n"
 "        <property name=\"Menu\" type=\"o\" access=\"read\" />\n"
+"        <property name=\"ItemIsMenu\" type=\"b\" access=\"read\" />\n"
 "        <property name=\"XAyatanaLabel\" type=\"s\" access=\"read\" />\n"
 "        <property name=\"XAyatanaLabelGuide\" type=\"s\" access=\"read\" />\n"
 "        <property name=\"XAyatanaOrderingIndex\" type=\"u\" access=\"read\" />\n"
 "        <property name=\"ToolTip\" type=\"(sa(iiay)ss)\" access=\"read\" />\n"
